File: libs/adsp-service-net-sdk/AdspId.cs
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using Adsp.Sdk.Utils;

namespace Adsp.Sdk;

[JsonConverter(typeof(AdspIdJsonConverter))]
public class AdspId : IEquatable<AdspId>
{
  private static readonly Regex UrnRegex = new(
    "^(?i:urn):ads(?<namespace>:[a-zA-Z0-9-]{1,50})(?<service>:[a-zA-Z0-9-]{1,50})?(?<api>:[a-zA-Z0-9-]{1,50})?(?<resource>:[a-zA-Z0-9-_/ ]{1,1000})?$",
    RegexOptions.Singleline | RegexOptions.CultureInvariant
  );

  public static AdspId Parse(string? urn)
  {
    if (String.IsNullOrEmpty(urn))
    {
      throw new ArgumentNullException(nameof(urn));
    }

    var match = UrnRegex.Match(urn);
    var @namespace = match.Groups["namespace"].Value.TrimStart(':');
    var service = match.Groups["service"].Value.TrimStart(':');
    var api = match.Groups["api"].Value.TrimStart(':');
    var resource = match.Groups["resource"].Value.TrimStart(':');

    ResourceType? type = null;
    if (!String.IsNullOrEmpty(resource))
    {
      type = ResourceType.Resource;
    }
    else if (!String.IsNullOrEmpty(api))
    {
      type = ResourceType.Api;
    }
    else if (!String.IsNullOrEmpty(service))
    {
      type = ResourceType.Service;
    }
    else if (!String.IsNullOrEmpty(@namespace))
    {
      type = ResourceType.Namespace;
    }

    if (type == null || (type == ResourceType.Resource && String.IsNullOrEmpty(api)))
    {
      throw new ArgumentException($"Specified urn '{urn}' is not an ADSP ID.");
    }

    return new AdspId((ResourceType)type, @namespace, service, api, resource);
  }

  public static AdspId Parse(Uri? urn)
  {
    return Parse(urn?.ToString());
  }

  private readonly ResourceType _type;
  public ResourceType Type
  {
    get { return _type; }
  }

  private readonly string _namespace;
  public string Namespace
  {
    get { return _namespace; }
  }

  private readonly string? _service;
  public string? Service
  {
    get { return _service; }
  }

  private readonly string? _api;
  public string? Api
  {
    get { return _api; }
  }

  private readonly string? _resource;
  public string? Resource
  {
    get { return _resource; }
  }

  private AdspId(ResourceType type, string @namespace, string? service, string? api, string? resource)
  {
    _type = type;
    _namespace = @namespace;
    _service = service;
    _api = api;
    _resource = resource;
  }

  public override string ToString()
  {
    return $"urn:ads:{Namespace}" +
      $"{(!String.IsNullOrEmpty(Service) ? $":{Service}" : "")}" +
      $"{(!String.IsNullOrEmpty(Api) ? $":{Api}" : "")}" +
      $"{(!String.IsNullOrEmpty(Resource) ? $":{Resource}" : "")}";
  }

  public bool Equals(AdspId? other)
  {
    return other != null && (Type, Namespace, Service, Api, Resource).Equals(
      (other.Type, other.Namespace, other.Service, other.Api, other.Resource)
    );
  }

  public override bool Equals(object? obj)
  {
    if (obj is AdspId other)
    {
      return Equals(other);
    }
    else
    {
      return base.Equals(obj);
    }
  }

  public override int GetHashCode()
  {
    return HashCode.Combine(Type, Namespace, Service, Api, Resource);
  }
}

File: libs/adsp-service-spring-sdk/src/main/java/ca/ab/gov/alberta/adsp/sdk/AdspId.java
package ca.ab.gov.alberta.adsp.sdk;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.springframework.util.Assert;
import org.springframework.util.ObjectUtils;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import ca.ab.gov.alberta.adsp.sdk.utils.AdspIdDeserializer;

@JsonDeserialize(using = AdspIdDeserializer.class)
public final class AdspId {
  private static final Pattern UrnPattern = Pattern.compile(
      "^(?i:urn):ads(?<namespace>:[a-zA-Z0-9-]{1,50})(?<service>:[a-zA-Z0-9-]{1,50})?(?<api>:[a-zA-Z0-9-]{1,50})?(?<resource>:[a-zA-Z0-9-_/ ]{1,1000})?$");

  public static AdspId parse(String urn) {
    Assert.hasLength(urn, "urn cannot be null or empty.");

    Matcher matcher = UrnPattern.matcher(urn);

    if (!matcher.find()) {
      throw new IllegalArgumentException("Specified urn is not an ADSP ID.");
    }

    var namespace = StringUtils.strip(matcher.group("namespace"), ":");
    var service = StringUtils.strip(matcher.group("service"), ":");
    var api = StringUtils.strip(matcher.group("api"), ":");
    var resource = StringUtils.strip(matcher.group("resource"), ":");

    ResourceType type = null;
    if (!ObjectUtils.isEmpty(resource)) {
      type = ResourceType.Resource;
    } else if (!ObjectUtils.isEmpty(api)) {
      type = ResourceType.Api;
    } else if (!ObjectUtils.isEmpty(service)) {
      type = ResourceType.Service;
    } else if (!ObjectUtils.isEmpty(namespace)) {
      type = ResourceType.Namespace;
    }
    return new AdspId(type, namespace, service, api, resource);
  }

  private final ResourceType type;
  private final String namespace;
  private final String service;
  private final String api;
  private final String resource;

  public ResourceType getType() {
    return type;
  }

  public String getNamespace() {
    return namespace;
  }

  public String getService() {
    return service;
  }

  public String getApi() {
    return api;
  }

  public String getResource() {
    return resource;
  }

  private AdspId(ResourceType type, String namespace, String service, String api, String resource) {
    this.type = type;
    this.namespace = namespace;
    this.service = service;
    this.api = api;
    this.resource = resource;
  }

  @Override
  public String toString() {
    return Stream.of("urn", "ads", namespace, service, api, resource)
        .filter((item) -> !StringUtils.isEmpty(item))
        .collect(Collectors.joining(":"));
  }

  @Override
  public boolean equals(Object other) {

    if (!(other instanceof AdspId)) {
      return false;
    }

    var otherId = (AdspId) other;

    return this.type.equals(otherId.type)
        && StringUtils.equals(this.namespace, otherId.namespace)
        && StringUtils.equals(this.service, otherId.service)
        && StringUtils.equals(this.api, otherId.api)
        && StringUtils.equals(this.resource, otherId.resource);
  }

  @Override
  public int hashCode() {
    return new HashCodeBuilder(3, 17)
        .append(this.type)
        .append(this.namespace)
        .append(this.service)
        .append(this.api)
        .append(this.resource)
        .toHashCode();
  }
}
